File: 08._layered_architecture/example-kotlin/src/main/kotlin/com/example/notes/domain/Note.kt
package com.example.notes.domain

import kotlinx.serialization.Serializable

@Serializable
data class Note(val id: Long, val title: String, val body: String)

class InvalidNoteException(message: String) : RuntimeException(message)

object NoteRules {
    const val MAX_TITLE_LENGTH = 100

    fun validateTitle(raw: String): String {
        val title = raw.trim()
        if (title.isEmpty()) throw InvalidNoteException("title cannot be blank")
        if (title.length > MAX_TITLE_LENGTH) {
            throw InvalidNoteException("title cannot exceed $MAX_TITLE_LENGTH characters")
        }
        return title
    }
}

File: 08._layered_architecture/example-python/notes/domain/note.py
from dataclasses import dataclass

MAX_TITLE_LENGTH = 100


class InvalidNoteError(Exception):
    pass


@dataclass(frozen=True)
class Note:
    id: int
    title: str
    body: str


def validate_title(raw: str) -> str:
    title = raw.strip()
    if not title:
        raise InvalidNoteError("title cannot be blank")
    if len(title) > MAX_TITLE_LENGTH:
        raise InvalidNoteError(f"title cannot exceed {MAX_TITLE_LENGTH} characters")
    return title
